packages/yarnpkg-core/sources/Report.ts

```diff
@@ -1,16 +1,36 @@
+import camelcase       from 'camelcase';
 import {PassThrough}   from 'stream';
 import {StringDecoder} from 'string_decoder';
 
 import {MessageName}   from './MessageName';
 import {Locator}       from './types';
 
+export type ReportErrorOptions = {
+  includeStack: boolean;
+  reportExtra: (report: Report) => void
+};
+
 export class ReportError extends Error {
   public reportCode: MessageName;
 
-  constructor(code: MessageName, message: string, public reportExtra?: (report: Report) => void) {
+  public reportExtra?: (report: Report) => void;
+
+  constructor(code: MessageName, message: string, {includeStack = true, reportExtra}: Partial<ReportErrorOptions> = {}) {
     super(message);
 
+    if (code !== MessageName.UNNAMED && code !== MessageName.EXCEPTION) {
+      this.name = camelcase(MessageName[code], {pascalCase: true});
+      if (!this.name.endsWith(`Error`) && !this.name.endsWith(`Exception`)) {
+        this.name += `Error`;
+      }
+    }
+
     this.reportCode = code;
+    this.reportExtra = reportExtra;
+
+    if (!includeStack) {
+      delete this.stack;
+    }
```

Comment:
Why do you need to delete the stack? It should already be hidden by the report functions, right?

Reply:
🤔 Indeed it should be but isn't for some reason (you can see it by deleting that part of the code and running `YARN_NPM_REGISTRY_SERVER=https://npm.pkg.github.com/download add left-pad` without auth).

Reply:
Found the reason: In `add` and `up` we were using `this.cli.error` to print the pretty error instead of `report.reportExceptionOnce`, which caused the stack to be preserved (my code 😬). I fixed it and it works correctly now:
![image](https://user-images.githubusercontent.com/32596136/106197001-d036dd80-61ba-11eb-94da-61dc6d76ca44.png)